File: server/src/models/DryingRequest.js
const mongoose = require('mongoose');

const DRYING_REQUEST_STATUSES = ['PENDING', 'ACCEPTED', 'REJECTED', 'CANCELLED', 'COMPLETED'];

const dryingRequestSchema = new mongoose.Schema(
  {
    householdId: { type: mongoose.Schema.Types.ObjectId, ref: 'Household', required: true },
    // Optional link to the wash turn that produced the clothes. Drying is a
    // separate responsibility from the machine lifecycle (see drying.service.js) —
    // this is informational only and never gates any drying transition.
    turnId: { type: mongoose.Schema.Types.ObjectId, ref: 'Turn', default: null },
    requesterId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true }, // clothes owner
    helperId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true }, // requested helper
    status: { type: String, enum: DRYING_REQUEST_STATUSES, default: 'PENDING' },
    acceptedAt: { type: Date, default: null },
    completedAt: { type: Date, default: null },
    rejectedAt: { type: Date, default: null },
    cancelledAt: { type: Date, default: null },
  },
  { timestamps: true }
);

dryingRequestSchema.index({ householdId: 1, status: 1 });
dryingRequestSchema.index({ helperId: 1, status: 1 });
dryingRequestSchema.index({ requesterId: 1, status: 1 });

// At most one active (PENDING/ACCEPTED) drying request per (requester, helper,
// turn). A task is identified by the wash turn it's associated with, or — for
// ad hoc requests with no turn — by the (requester, helper) pair itself: Mongo's
// partial index treats null turnId as an equal value across documents, so a
// second ad hoc request between the same pair can't be created while one is
// already active. Once a request leaves PENDING/ACCEPTED it drops out of this
// partial filter, so a brand new request for a later task is never blocked.
dryingRequestSchema.index(
  { requesterId: 1, helperId: 1, turnId: 1 },
  { unique: true, partialFilterExpression: { status: { $in: ['PENDING', 'ACCEPTED'] } } }
);

module.exports = mongoose.model('DryingRequest', dryingRequestSchema);
module.exports.DRYING_REQUEST_STATUSES = DRYING_REQUEST_STATUSES;

File: app/lib/core/models/drying_request.dart
enum DryingRequestStatus { pending, accepted, rejected, cancelled, completed }

DryingRequestStatus _parseDryingStatus(String s) => switch (s) {
      'PENDING' => DryingRequestStatus.pending,
      'ACCEPTED' => DryingRequestStatus.accepted,
      'REJECTED' => DryingRequestStatus.rejected,
      'CANCELLED' => DryingRequestStatus.cancelled,
      'COMPLETED' => DryingRequestStatus.completed,
      _ => DryingRequestStatus.pending,
    };

class DryingRequest {
  final String id;
  final String householdId;
  final String? turnId;
  final String requesterId; // clothes owner
  final String helperId; // asked to dry them
  final DryingRequestStatus status;
  final DateTime createdAt;
  final DateTime? acceptedAt;
  final DateTime? completedAt;
  final DateTime? rejectedAt;
  final DateTime? cancelledAt;

  DryingRequest({
    required this.id,
    required this.householdId,
    this.turnId,
    required this.requesterId,
    required this.helperId,
    required this.status,
    required this.createdAt,
    this.acceptedAt,
    this.completedAt,
    this.rejectedAt,
    this.cancelledAt,
  });

  bool get isPending => status == DryingRequestStatus.pending;
  bool get isAccepted => status == DryingRequestStatus.accepted;

  factory DryingRequest.fromJson(Map<String, dynamic> json) => DryingRequest(
        id: json['_id'] as String,
        householdId: json['householdId'] as String,
        turnId: json['turnId'] as String?,
        requesterId: json['requesterId'] as String,
        helperId: json['helperId'] as String,
        status: _parseDryingStatus(json['status'] as String),
        createdAt: DateTime.parse(json['createdAt'] as String),
        acceptedAt: json['acceptedAt'] != null ? DateTime.parse(json['acceptedAt'] as String) : null,
        completedAt: json['completedAt'] != null ? DateTime.parse(json['completedAt'] as String) : null,
        rejectedAt: json['rejectedAt'] != null ? DateTime.parse(json['rejectedAt'] as String) : null,
        cancelledAt: json['cancelledAt'] != null ? DateTime.parse(json['cancelledAt'] as String) : null,
      );
}

/// A directional drying-favor balance between the current user and one other
/// household member. Never both nonzero at once — see server drying.service.js.
class DryingFavorBalance {
  final String userId;
  final int youOwe;
  final int owesYou;

  DryingFavorBalance({required this.userId, required this.youOwe, required this.owesYou});

  factory DryingFavorBalance.fromJson(Map<String, dynamic> json) => DryingFavorBalance(
        userId: json['userId'] as String,
        youOwe: json['youOwe'] as int,
        owesYou: json['owesYou'] as int,
      );
}
